src/server/ua_services_nodemanagement.c

```diff
@@ -754,6 +778,33 @@ addInterfaceChildren(UA_Server *server, UA_Session *session,
             break;
     }
 
+    if (retval == UA_STATUSCODE_GOOD) {
+        for (size_t i = 0; i < hierarchySize; ++i) {
+            UA_ExpandedNodeId targetId;
+            UA_ExpandedNodeId_init(&targetId);
+            targetId.nodeId = hierarchy[i];
+
+            /*
+             * This is bad, some other thread could sneak in here.
+             * Unfortunately, there is no other choice due to the lock assert.
+             */
+            UA_UNLOCK(&server->serviceMutex);
+            retval = UA_Server_addReference(server, head->nodeId,
```

Comment:
There is an internal version that does not take locks.
You might have to expose it in ua_server_internal.h.